File: ts/project/.sdk/tm/py/feature/audit_feature.py
# ProjectName SDK audit feature

from __future__ import annotations
import time

from feature.base_feature import ProjectNameBaseFeature


# Audit trail. Emits a structured record for every operation — who (actor),
# what (entity + op), the outcome, and a correlation id — suitable for
# compliance logging. Records accumulate on `client._audit["records"]`
# (bounded by `max`, default 1000) and, when a `sink` callback is supplied,
# are also pushed to it (e.g. to forward to a SIEM). The actor is taken
# from a per-call `ctrl` actor, then options (`actor`), then 'anonymous'.
# Timestamps use the injectable `now` clock so tests stay deterministic.
class ProjectNameAuditFeature(ProjectNameBaseFeature):
    def __init__(self):
        super().__init__()
        self.version = "0.0.1"
        self.name = "audit"
        self.active = True
        self.client = None
        self.options = {}
        self.seq = 0

    def init(self, ctx, options):
        self.client = ctx.client
        self.options = options if isinstance(options, dict) else {}

        if self.options.get("active") is True:
            self.active = True
        else:
            self.active = False

        self.seq = 0

        client = self.client
        if getattr(client, "_audit", None) is None:
            client._audit = {"records": []}

    def PreDone(self, ctx):
        if not self.active:
            return
        # Outcome reflects the actual result; a non-2xx reaches PreDone
        # before the pipeline raises.
        result = ctx.result
        ok = result is not None and result.ok is not False and result.err is None
        self._emit(ctx, "ok" if ok else "error")

    def PreUnexpected(self, ctx):
        if not self.active:
            return
        self._emit(ctx, "error")

    def _emit(self, ctx, outcome):
        # One record per operation (PreDone + a following PreUnexpected on a
        # non-2xx must not double-log).
        if getattr(ctx, "_audit_seen", False):
            return
        ctx._audit_seen = True

        self.seq += 1

        actor = getattr(ctx.ctrl, "actor", None) if ctx.ctrl is not None else None
        if actor is None:
            actor = self.options.get("actor")
        if actor is None:
            actor = "anonymous"

        record = {
            "seq": self.seq,
            "ts": self._now(),
            "actor": actor,
            "entity": (ctx.op.entity if ctx.op is not None else None) or "_",
            "op": (ctx.op.name if ctx.op is not None else None) or "_",
            "outcome": outcome,
            "status": ctx.result.status if ctx.result is not None else None,
            "correlationId": ctx.id,
        }

        records = self.client._audit["records"]
        records.append(record)
        mx = 1000 if self.options.get("max") is None else self.options.get("max")
        while len(records) > mx:
            records.pop(0)

        sink = self.options.get("sink")
        if callable(sink):
            try:
                sink(record)
            except Exception:
                pass

    def _now(self):
        now = self.options.get("now")
        if callable(now):
            return now()
        return time.time() * 1000

File: ts/project/.sdk/tm/rb/feature/audit_feature.rb
# ProjectName SDK audit feature
#
# Audit trail. Emits a structured record for every operation - who (actor),
# what (entity + op), the outcome, and a correlation id - suitable for
# compliance logging. Records accumulate on the client (bounded by "max",
# default 1000) and, when a "sink" callback is supplied, are also pushed to
# it (e.g. to forward to a SIEM). The actor is taken from a per-call
# ctrl actor, then options "actor", then "anonymous". Timestamps use the
# injectable "now" clock (ms) so tests stay deterministic. One record per
# operation (emit-once).

require_relative 'base_feature'

class ProjectNameAuditFeature < ProjectNameBaseFeature
  def initialize
    super
    @version = "0.0.1"
    @name = "audit"
    # Inactive until init (feature_init only fires init when active).
    @active = false
    @client = nil
    @options = {}
    @seq = 0
    @seen = {}
  end

  def init(ctx, options)
    @client = ctx.client
    @options = options.is_a?(Hash) ? options : {}
    @active = @options["active"] == true
    @seq = 0
    @seen = {}

    return unless @active

    if @client.instance_variable_get(:@_audit).nil?
      @client.instance_variable_set(:@_audit, { "records" => [] })
    end
  end

  def PreDone(ctx)
    return unless @active
    # Outcome reflects the actual result: a non-2xx reaches PreDone before
    # the pipeline raises.
    ok = !ctx.result.nil? && ctx.result.ok == true && ctx.result.err.nil?
    _emit(ctx, ok ? "ok" : "error")
  end

  def PreUnexpected(ctx)
    return unless @active
    _emit(ctx, "error")
  end

  def _emit(ctx, outcome)
    # One record per operation (PreDone + a following PreUnexpected on a
    # non-2xx must not double-log).
    return if @seen.key?(ctx)
    @seen[ctx] = true
    @seq += 1

    record = {
      "seq" => @seq,
      "ts" => _now,
      "actor" => _actor(ctx),
      "entity" => ctx.op ? ctx.op.entity : "_",
      "op" => ctx.op ? ctx.op.name : "_",
      "outcome" => outcome,
      "status" => ctx.result ? ctx.result.status : nil,
      "correlationId" => ctx.id,
    }

    track = @client.instance_variable_get(:@_audit)
    if track.nil?
      track = { "records" => [] }
      @client.instance_variable_set(:@_audit, track)
    end
    recs = track["records"]
    recs << record
    max = @options["max"].nil? ? 1000 : @options["max"]
    recs.shift while recs.length > max

    sink = @options["sink"]
    if sink.is_a?(Proc)
      begin
        sink.call(record)
      rescue StandardError
      end
    end
  end

  def _actor(ctx)
    if ctx.ctrl && ctx.ctrl.respond_to?(:actor) && !ctx.ctrl.actor.nil?
      return ctx.ctrl.actor
    end
    @options["actor"] || "anonymous"
  end

  def _now
    now = @options["now"]
    return now.call if now.is_a?(Proc)
    (Time.now.to_f * 1000).to_i
  end
end
